File: lib-segment/src/test/java/websiteschema/mpsegment/performance/AccuracyTest.java
package websiteschema.mpsegment.performance;

import junit.framework.Assert;
import org.junit.Test;
import websiteschema.mpsegment.core.SegmentEngine;
import websiteschema.mpsegment.core.SegmentWorker;
import websiteschema.mpsegment.tools.accurary.ErrorAnalyzer;
import websiteschema.mpsegment.tools.accurary.SegmentAccuracy;
import websiteschema.mpsegment.tools.accurary.SegmentErrorType;

import java.io.IOException;
import java.util.Map;

public class AccuracyTest {

    @Test
    public void should_be_higher_than_93_percent_with_segment_minimum() throws IOException {
        SegmentWorker segmentWorker =
                SegmentEngine.getInstance().getSegmentWorker(
                        "separate.xingming -> true",
                        "minimize.word -> true"
                );
        SegmentAccuracy segmentAccuracy = new SegmentAccuracy("PFR-199801-utf-8.txt", segmentWorker);
        segmentAccuracy.checkSegmentAccuracy();
        System.out.println("Accuracy rate of segment is: " + segmentAccuracy.getAccuracyRate());
        System.out.println("There are " + segmentAccuracy.getWrong() + " errors and total expect word is " + segmentAccuracy.getTotalWords() + " when doing accuracy test.");

        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord).getErrorOccurTimes() + " errors because of new word.");
        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.NER_NR).getErrorOccurTimes() + " errors because of name recognition.");
        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.NER_NS).getErrorOccurTimes() + " errors because of place name recognition.");
        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate).getErrorOccurTimes() + " errors because of contain disambiguate.");
        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.Other).getErrorOccurTimes() + " other errors");

        System.out.println("Possible " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord).getWords().size() + " new words");
        System.out.println("Total count: " + getWordsCount(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord))+", they are:");
        System.out.println(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord).getWords());
        System.out.println("Those " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate).getWords().size() + " words maybe could delete from dictionary");
        System.out.println("Total count: " + getWordsCount(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate)) + ", they are:");
        System.out.println(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate).getWords());

        Assert.assertTrue(segmentAccuracy.getAccuracyRate() > 0.94021 * 0.99);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord).getErrorOccurTimes() <= 23891 * 1.05);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.NER_NR).getErrorOccurTimes() <= 4156 * 1.05);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.NER_NS).getErrorOccurTimes() <= 3282 * 1.05);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate).getErrorOccurTimes() <= 35317 * 1.05);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.Other).getErrorOccurTimes() <= 3610 * 1.05);
    }

    @Test
    public void should_be_higher_than_93_percent() throws IOException {
        SegmentWorker segmentWorker = SegmentEngine.getInstance().getSegmentWorker("separate.xingming -> true");
        SegmentAccuracy segmentAccuracy = new SegmentAccuracy("PFR-199801-utf-8.txt", segmentWorker);
        segmentAccuracy.checkSegmentAccuracy();
        System.out.println("Accuracy rate of segment is: " + segmentAccuracy.getAccuracyRate());
        System.out.println("There are " + segmentAccuracy.getWrong() + " errors and total expect word is " + segmentAccuracy.getTotalWords() + " when doing accuracy test.");

        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord).getErrorOccurTimes() + " errors because of new word.");
        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.NER_NR).getErrorOccurTimes() + " errors because of name recognition.");
        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.NER_NS).getErrorOccurTimes() + " errors because of place name recognition.");
        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate).getErrorOccurTimes() + " errors because of contain disambiguate.");
        System.out.println("There are " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.Other).getErrorOccurTimes() + " other errors");

        System.out.println("Possible " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord).getWords().size() + " new words");
        System.out.println("Total count: " + getWordsCount(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord))+", they are:");
        System.out.println(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord).getWords());
        System.out.println("Those " + segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate).getWords().size() + " words maybe could delete from dictionary");
        System.out.println("Total count: " + getWordsCount(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate)) + ", they are:");
        System.out.println(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate).getWords());

        Assert.assertTrue(segmentAccuracy.getAccuracyRate() > 0.93994 * 0.99);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.UnknownWord).getErrorOccurTimes() <= 23534 * 1.05);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.NER_NR).getErrorOccurTimes() <= 4155 * 1.05);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.NER_NS).getErrorOccurTimes() <= 3036 * 1.05);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.ContainDisambiguate).getErrorOccurTimes() <= 36202 * 1.05);
        Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(SegmentErrorType.Other).getErrorOccurTimes() <= 3612 * 1.05);
    }

    private int getWordsCount(ErrorAnalyzer errorAnalyzer) {
        Map<String,Integer> words = errorAnalyzer.getWords();
        int count = 0;
        for(String word : words.keySet()) {
            count += words.get(word);
        }
        return count;
    }
}

File: lib-segment-scala/src/test/scala/websiteschema/mpsegment/performance/AccuracyTest.scala
package websiteschema.mpsegment.performance

import junit.framework.Assert
import org.junit.Test
import websiteschema.mpsegment.core.SegmentWorker
import websiteschema.mpsegment.tools.accurary.{NerNameStatisticData, ErrorAnalyzer, SegmentAccuracy}
import websiteschema.mpsegment.tools.accurary.SegmentErrorType._

class AccuracyTest {

  val debug = false

  @Test
  def should_be_higher_than_93_percent_with_segment_minimum() {
    val segmentWorker = SegmentWorker(
        "separate.xingming -> true",
//        "segment.bigram -> word-bigram.dat",
        "minimize.word -> true"
      )
    val segmentAccuracy = new SegmentAccuracy("PFR-199801-utf-8.txt", segmentWorker)
    segmentAccuracy.checkSegmentAccuracy()
    println("Accuracy rate of segment is: " + segmentAccuracy.getAccuracyRate())
    println("There are " + segmentAccuracy.getWrong() + " errors and total expect word is " + segmentAccuracy.getTotalWords() + " when doing accuracy test.")

    println("There are " + segmentAccuracy.getErrorAnalyzer(UnknownWord).getErrorOccurTimes() + " errors because of new word.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(NER_NR).getErrorOccurTimes() + " errors because of name recognition.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(NER_NS).getErrorOccurTimes() + " errors because of place name recognition.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(ContainDisambiguate).getErrorOccurTimes() + " errors because of contain disambiguate.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(Other).getErrorOccurTimes() + " other errors")

    printDetails(segmentAccuracy)

    Assert.assertTrue(segmentAccuracy.getAccuracyRate() > 0.94021 * 0.99)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(UnknownWord).getErrorOccurTimes() <= 23000 * 1.05)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(NER_NR).getErrorOccurTimes() <= 3000 * 1.05)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(NER_NS).getErrorOccurTimes() <= 3200 * 1.05)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(ContainDisambiguate).getErrorOccurTimes() <= 33000 * 1.05)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(Other).getErrorOccurTimes() <= 3800 * 1.05)
  }

  @Test
  def should_recognize_all_the_name_entities() {
    val segmentWorker = SegmentWorker(
        "separate.xingming -> true",
        "minimize.word -> true"
      )

    val segmentAccuracy = new SegmentAccuracy("test-pfr-corpus.txt", segmentWorker)
    segmentAccuracy.checkSegmentAccuracy()
    println("Accuracy rate of segment is: " + segmentAccuracy.getAccuracyRate())
    println("There are " + segmentAccuracy.getWrong() + " errors and total expect word is " + segmentAccuracy.getTotalWords() + " when doing accuracy test.")

    println("There are " + segmentAccuracy.getErrorAnalyzer(UnknownWord).getErrorOccurTimes() + " errors because of new word.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(NER_NR).getErrorOccurTimes() + " errors because of name recognition.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(NER_NS).getErrorOccurTimes() + " errors because of place name recognition.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(ContainDisambiguate).getErrorOccurTimes() + " errors because of contain disambiguate.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(Other).getErrorOccurTimes() + " other errors")

    printDetails(segmentAccuracy)

    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(NER_NR).getErrorOccurTimes() <= 4)
  }

  @Test
  def should_be_higher_than_93_percent() {
    val segmentWorker = SegmentWorker("separate.xingming -> true")
    val segmentAccuracy = new SegmentAccuracy("PFR-199801-utf-8.txt", segmentWorker)
    segmentAccuracy.checkSegmentAccuracy()
    println("Accuracy rate of segment is: " + segmentAccuracy.getAccuracyRate())
    println("There are " + segmentAccuracy.getWrong() + " errors and total expect word is " + segmentAccuracy.getTotalWords() + " when doing accuracy test.")

    println("There are " + segmentAccuracy.getErrorAnalyzer(UnknownWord).getErrorOccurTimes() + " errors because of new word.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(NER_NR).getErrorOccurTimes() + " errors because of name recognition.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(NER_NS).getErrorOccurTimes() + " errors because of place name recognition.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(ContainDisambiguate).getErrorOccurTimes() + " errors because of contain disambiguate.")
    println("There are " + segmentAccuracy.getErrorAnalyzer(Other).getErrorOccurTimes() + " other errors")

    printDetails(segmentAccuracy)

    Assert.assertTrue(segmentAccuracy.getAccuracyRate() > 0.94000 * 0.99)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(UnknownWord).getErrorOccurTimes() <= 23000 * 1.05)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(NER_NR).getErrorOccurTimes() <= 3000 * 1.05)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(NER_NS).getErrorOccurTimes() <= 3200 * 1.05)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(ContainDisambiguate).getErrorOccurTimes() <= 34000 * 1.05)
    Assert.assertTrue(segmentAccuracy.getErrorAnalyzer(Other).getErrorOccurTimes() <= 3800 * 1.05)
  }

  private def printDetails(segmentAccuracy: SegmentAccuracy) {
    if (debug) {
      println("Possible " + segmentAccuracy.getErrorAnalyzer(UnknownWord).getWords().size + " new words")
      println("Total count: " + getWordsCount(segmentAccuracy.getErrorAnalyzer(UnknownWord)) + ", they are:")
      println(segmentAccuracy.getErrorAnalyzer(UnknownWord).getWords())
      println("Those " + segmentAccuracy.getErrorAnalyzer(ContainDisambiguate).getWords().size + " words maybe could delete from dictionary")
      println("Total count: " + getWordsCount(segmentAccuracy.getErrorAnalyzer(ContainDisambiguate)) + ", they are:")
      println(segmentAccuracy.getErrorAnalyzer(ContainDisambiguate).getWords())
    }
    NerNameStatisticData.print
  }

  private def getWordsCount(errorAnalyzer: ErrorAnalyzer): Int = {
    val words = errorAnalyzer.getWords()
    var count = 0
    for (word <- words.keys) {
      count += words(word)
    }
    return count
  }
}
